Happi Quality-of-Life changes

unpairedbracket opened the pull request:
Some minor changes to happi to improve the user experience somewhat (and small bug fixes for some probe diagnostics and the transparent option):
* 2D (image) plots can accept the `norm` kwarg now, enabling log plots with nice colorbars along with more unusual normalisations like power laws and symmetric log
* An argument `vsym` that lets the range of a color axis vary automatically according to the data being plotted (e.g. in an animation) but forces it to remain centred on zero
* Nicer behaviour of colorbars: they now fit to the height of the plot instead of changing their relative size as the window changes size, have settable labels (useful for multiPlot) and width/padding. Default padding isn't currently auto-adjusted for multiPlot instances to stop overlab of one colorbar with the other's label, though this may be doable if desired.
* Arbitrary callable `data_transform` on Diagnostics, which can e.g. apply a convolution kernel such as a Gaussian blur to data before plotting.

and the bug fixes:
* 3D Probes didn't behave properly when sliced over axis1 or axis2
* `transparent` option to multiPlot caused an error if the colormap in use wasn't a `LinearSegmentedColormap`; now works for any colormap class which defines a `__copy__()` method.

mccoys commented on doc/Sphinx/post-processing.rst:
```diff
@@ -98,6 +98,8 @@ Open a Scalar diagnostic
   * ``data_log``:
      | If ``True``, then :math:`\log_{10}` is applied to the output.
+  * ``data_transform``:
+     | If ``callable``, is applied to the output.
```
Using the word "callable" is a little technical. I always try to make the documentation as accessible as possible. Wouldn't it be better to say something like "a function applied to the output if defined" ?

unpairedbracket replied:
Good point - I'll change the documentation here to make it a bit easier to understand

mccoys commented on doc/Sphinx/post-processing.rst:
```diff
@@ -442,6 +444,9 @@ to manipulate the plotting options:
 
 * ``figure``: The figure number that is passed to matplotlib.
 * ``vmin``, ``vmax``: data value limits.
+* ``vsym``: indicates that data is symmetric about zero. In the absence
```
I wonder whether it is possible to make a better parameter `vsym`. My point is that we need to set this parameter to `True` then choose a value for `vmax` if we want to set the value to a fixed number. What if `vsym` completely replaces `vmin` and `vmax` ? It could be set to `True` for auto limits, or to a float number for fixed limits `-vsym, vsym`. This is just an idea.

Anyways, the current documentation is a little confusing because it does not state that this parameter must be set to `True`. Furthermore, it does not explain the behavior when `vmax` and/or `vmin` are set to some value.

unpairedbracket replied:
Thanks for the suggestion, that's a really good idea. The way `vsym` currently works is that it only sets the data limits at all if both `vmin` and `vmax` are unset, with the reasoning being that if the user wants fixed symmetric limits it's easy to set that manually. I agree it would be a lot nicer to be able to set fixed limits directly in the `vsym` parameter, so I'll implement a change to allow this.

Do you have an idea of what the behaviour should be if `vsym` is set to a float and also one or both of `vmin`/`vmax` are set? I'd probably emit a warning and override the `vsym` limits with what's set in `vmin` and/or `vmax`, but if you think a different solution would be better I'll implement it that way.

mccoys replied:
I somewhat believe `vsym` should take precedence as it is only a single instruction. This is not very important anyways. I would do something like:
```python
if vsym:
    if vmin or vmax:
        print("Some warning ...")
    if vsym is True:
        vmax = abs(A).max()
    else:
        vmax = abs(vsym)
    vmin = -vmax
```

mccoys reviewed (commented):
Thank you very much for this contribution ! I have written a couple of comments inline, but I am also wondering whether you tested this with both python 2 and 3. Furthermore, which version of matplotlib did you try ?

unpairedbracket commented:
These changes were written with python version 3.7.3 and matplotlib version 3.1.3. I didn't think to test with python 2 versions as it's end-of-life now, but of course some systems will still be working with python 2 so I'll see if I can test with a python 2 environment. I couldn't find a listed minimum version of matplotlib that happi targets, but I can see that `axes_grid1` seems to have existed since v2.0.1 and I think everything else I've used has existed for at least that long as well. If you have a minimum matplotlib version that you try to support I can try to make sure it works there as well.

mccoys commented:
Concerning the version for matplotlib, we do not really set a version, but we will fix things when we encounter an issue with an older version (it is quite rare).

For python2, it is more important as many older machines do not have python3.

unpairedbracket commented:
I've changed the behavior of the `vsym` parameter as suggested in your comment, updated the documentation to match, and also clarified the documentation for the `data_transform` parameter. It now reads "If this is set to a function, the function is applied to the output before plotting." and the documentation for `vsym` reads "indicates that data is symmetric about zero. If set to a number, data limits are set to `[-vsym, vsym]`. If set to `True`, limits autoscale to the values present in the simulation, but remain symmetric. If `vmin` or `vmax` are set, they are ignored." I hope that covers everything

mccoys commented:
Thank you very much for your efforts !
